sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/QueryTransform.java

```diff
@@ -78,25 +76,18 @@
     }
   }
 
-  private PCollectionTuple toPCollectionTuple(PInput inputs) {
-    return (inputs instanceof PCollection)
-        ? PCollectionTuple.of(new TupleTag<>(PCOLLECTION_NAME), toRows(inputs))
-        : tupleOfAllInputs(inputs.getPipeline(), inputs.expand());
-  }
-
-  private PCollectionTuple tupleOfAllInputs(
-      Pipeline pipeline,
-      Map<TupleTag<?>, PValue> taggedInputs) {
-
-    PCollectionTuple tuple = PCollectionTuple.empty(pipeline);
-
-    for (Map.Entry<TupleTag<?>, PValue> input : taggedInputs.entrySet()) {
-      tuple = tuple.and(
-          new TupleTag<>(input.getKey().getId()),
-          toRows(input.getValue()));
+  private BeamSqlTableProvider toTableProvider(PInput inputs) {
+    ImmutableMap.Builder<String, BeamSqlTable> tables = ImmutableMap.builder();
+    if (inputs instanceof PCollection) {
+      tables.put(PCOLLECTION_NAME,
+          new BeamPCollectionTable(toRows(inputs)));
+    } else {
+      for (Map.Entry<TupleTag<?>, PValue> input : inputs.expand().entrySet()) {
+        tables.put(input.getKey().getId(),
+            new BeamPCollectionTable(toRows(input.getValue())));
+      }
```

Comment:
nit: I would avoid stateful if/else with loops with generics, hurts readability. Might consider extracting something like this:

```java
if (input instanceof PCollection) {
  return 
     ImmuableMap.of(
        PCOLLECTION_NAME, 
        new BeamPCollectionTable(toRows(inputs)))
}

return
    inputs
        .expand()
        .entrySet()
        .stream()
        .collect(
            toMap(
                keyedPCollection -> keyedPCollection.getKey().getId(),
                keyedPCollection -> keyedPCollection.getValue()))
```

and then create BeamSqlTableProvider outside

Reply:
Incidentally, here (or maybe it is nearby beneath a fold) seems like a good place to (possibly redundantly) explain that a PCollection makes a single magic table while any other kind of input uses `expand()` to make many tables using the tags as names.

Reply:
I agree this is hard to read, I disagree that java streams makes it more readable. I've restructured it like you've suggested otherwise and added the comment.